test/truster/README.md

```diff
@@ -0,0 +1,777 @@
+Yes, since the **data** parameter also turned out to be symbolic, we got into the same pattern again and got the same error. However, this time, knowing that we can get to a symbolic call of a symbolic address in any contract, we will use a more universal solution!
+## Global storage
+Let's create a library global storage contract that can be accessed from anywhere:
+```solidity
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.25;
+
+import "./halmos-cheatcodes/src/SymTest.sol";
+import {Test, console} from "forge-std/Test.sol";
+
+contract GlobalStorage is SymTest {
+    // uint256->address mapping to have an ability to iterate over addresses
+    mapping (uint256 => address) addresses;
+    mapping (address => string) names_by_addr;
+
+    uint256 addresses_list_size = 0;
+
+    // Addresses and names information is stored using this setter
+    function add_addr_name_pair (address addr, string memory name) public {
+        addresses[addresses_list_size] = addr;
+        addresses_list_size++;
+        names_by_addr[addr] = name;
+    }
+
+    /*
+    ** It is expected to receive a symbolic address as a parameter
+    ** This function should return some concrete address and corresponding data.
+    ** In the case of symbolic execution, the brute force over addresses
+    ** is happening here!
+    */
+    function get_concrete_from_symbolic (address /*symbolic*/ addr) public view 
+                                        returns (address ret, bytes memory data) 
+    {
+        for (uint256 i = 0; i < addresses_list_size; i++) {
+            if (addresses[i] == addr) {
+                string memory name = names_by_addr[addr];
+                return (addresses[i], svm.createCalldata(name));
+            }
+        }
+        revert(); // Ignore cases when addr is not some concrete known address
+    }
+}
+```
+
+We will not dwell on the implementation details of this contract. I will only say that this is a contract in which you can store address->contract name pairs. And also with its help you can conveniently brute force addresses symbolically. It is easier to show how to use it in practice. First, let's prepare Global Storage in Truster_Halmos.t.sol:
+```solidity
+...
+import "lib/GlobalStorage.sol";
+...
+contract TrusterChallenge is Test {
+...
+    GlobalStorage public glob; // Add global storage contract
+    DamnValuableToken public token;
+...
+    function setUp() public {
+    ...
+        // Deploy global storage. It'll have a "0xaaaa0002" address
+        glob = new GlobalStorage();
+    ...
+        glob.add_addr_name_pair(address(token), "DamnValuableToken");
+        glob.add_addr_name_pair(address(pool), "TrusterLenderPool");
+        vm.stopPrank();
+    }
+...
+```
+We will use it in SymbolicAttacker:
+```solidity
+...
+import "lib/GlobalStorage.sol";
+...
+contract SymbolicAttacker is Test, SymTest {
+    // We can hardcode this address for convenience
+    GlobalStorage glob = GlobalStorage(address(0xaaaa0002)); 
```

Comment:
can we pass it as a constructor parameter? hardcoding it seems brittle (for instance if we move things around a bit and addresses change, it would require debugging what's going on)

Reply:
Of course, we normally expect to pass such contracts through the constructor. However, I did this specifically to emphasize that we can use this GlobalStorage from anywhere in the application. Imagine that we needed it in a contract generated by some factory contract. If we pass them through the constructor - we will have to rewrite each of these contracts by adding GlobalStorage. I thought it was less convenient than just hardcoding it where it was needed. And logging in at the very beginning immediately gives us the address of this contract.
What do you think?

Reply:
This discussion will definitely end as soon as "native" storage appears in Halmos :). For example, Clang Static Analyzer also uses symbolic code execution and there is an possibility to use such storage to collect the necessary information. How about an idea for improvement?